crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3723,43 +3759,105 @@ impl StringPartsCollector {
+    // The reflected dunder has priority if the right-hand side is a strict subclass of the left-hand side.
+    if left_class != right_class && right_class.is_subclass_of(db, left_class) {
+        call_dunder(op.reflect(), right_class, left_class)
+            .or_else(|| call_dunder(op, left_class, right_class))
+    } else {
+        call_dunder(op, left_class, right_class)
+            .or_else(|| call_dunder(op.reflect(), right_class, left_class))
+    }
+    .or_else(|| {
+        // When no appropriate method returns any value other than NotImplemented,
+        // the `==` and `!=` operators will fall back to `is` and `is not`, respectively.
+        // refer to `<https://docs.python.org/3/reference/datamodel.html#object.__eq__>`
+        if matches!(op, RichCompareOperator::Eq | RichCompareOperator::Ne) {
+            Some(KnownClass::Bool.to_instance(db))
+        } else {
+            None
+        }
+    })
+    .ok_or_else(|| CompareUnsupportedError {
         op: op.into(),
-        left_ty: Type::Instance(left),
-        right_ty: Type::Instance(right),
+        left_ty: Type::Instance(left_class),
+        right_ty: Type::Instance(right_class),
     })
 }
 
+/// Performs a membership test (`in` and `not in`) between two instances and returns the resulting type.
+/// The behavior can be customized in Python by implementing `__contains__`, `__iter__`, or `__getitem__` methods.
+/// See `<https://docs.python.org/3/reference/datamodel.html#object.__contains__>`
+/// and `<https://docs.python.org/3/reference/expressions.html#membership-test-details>`
+fn perform_membership_test_comparison<'db>(
+    db: &'db dyn Db,
+    left_class: ClassType<'db>,
+    right_class: ClassType<'db>,
+    op: MembershipTestCompareOperator,
+) -> Result<Type<'db>, CompareUnsupportedError<'db>> {
+    let (left_instance, right_instance) = (Type::Instance(left_class), Type::Instance(right_class));
+
+    let contains_dunder = right_class.class_member(db, "__contains__");
+
+    let compare_result_opt = if contains_dunder.is_unbound() {
+        // If `__contains__` is unbound, the code falls back to checking the `__iter__` method for an iteration-based membership test.
+        right_class
+            .class_member(db, "__iter__")
+            .call(db, &[right_instance])
+            .return_ty(db)
+            .map(|_| KnownClass::Bool.to_instance(db))
+            .or_else(|| {
+                // If `__iter__` is not defined, the code falls back to the old-style iteration protocol,
+                // where Python iterates over natural numbers (0..) until an IndexError is raised or the item is found.
+                right_class
+                    .class_member(db, "__getitem__")
+                    .call(db, &[right_instance, KnownClass::Int.to_instance(db)]) // Simulate item access with an integer index.
+                    .return_ty(db)
+                    .map(|_item_ty| KnownClass::Bool.to_instance(db))
+            })
```

Comment:
can some of this be simplified by using our pre-existing `Type::iterate()` method?